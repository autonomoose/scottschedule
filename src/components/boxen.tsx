/*
    various boxes
*/
import React from 'react';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';

interface iCaptionBox {
    caption: string,
    color?: string,
    xpad?: string, // rem
}
// quick caption box
export const CaptionBox = (props: iCaptionBox) => {
    const { caption, color, xpad } = props;
    const wkColor = color || 'inherit';
    const wkXpad = xpad || '0.5rem';
    return (
      <Box px={wkXpad} mt={-1.75}>
        <Typography variant='caption' color={wkColor}
          sx={{ bgcolor:'background.paper', position: 'relative', zIndex: '2'}} >
          &nbsp;{caption}&nbsp;
        </Typography>
      </Box>
)}

