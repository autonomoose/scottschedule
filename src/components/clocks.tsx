// more clocks
//   - clocks must avoid hooks, clocks are conditionals
//   - define new clock and register it with widget

import React from 'react';

import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';

// use this as an outside interface to clocks
interface ClockWidgetProps {
  currClock: string,
  onComplete: (status: string) => void,
  currSched?: string,
  descrSched?: string,
}
export const ClockWidget = (props: ClockWidgetProps) => {
    return(
      <Box display='flex' flexDirection='column'>
      {
        {
          'digital1': <ClockDigital1 onComplete={props.onComplete} />,
          'digital2': <ClockDigital2 onComplete={props.onComplete} />,
        }[props.currClock]
      }
      </Box>
)}

interface ClockProps {
  onComplete: (status: string) => void,
}
export const ClockDigital2 = (props: ClockProps) => {
    return(
      <>
        <Button data-testid='change clock' sx={{margin: 0}} onClick={() => props.onComplete('scheduler')}>
          <Typography variant='h1' id='mainclock' data-testid='mainclock'
            sx={{fontSize:150, fontWeight: 600, padding: 0, margin: 0}}>00:00</Typography>
        </Button>
        <Typography variant='subtitle1' id='mainpm' data-testid='mainpm' sx={{fontSize:20, fontWeight: 600}}>
          PM
        </Typography>
        <Box mx={4} display='flex' justifyContent='space-between'>
          <Typography mx={1} variant='h4' id='maindate' data-testid='maindate'>Day, 01/01</Typography>
          <Button onClick={() => props.onComplete('close')}>Scheduler</Button>
        </Box>
      </>
) };

export const ClockDigital1 = (props: ClockProps) => {
    return(
      <>
        <Button data-testid='change clock' sx={{margin: 0, padding: 0}} onClick={() => props.onComplete('digital2')}>
          <Box display='flex' justifyContent='flex-start' alignItems='stretch'>
            <Box mt={1} display='flex' alignItems='flex-start'>
              <Typography id='mainpm' data-testid='mainpm'
               sx={{lineHeight:1, fontSize:20, fontWeight: 600}}>
                PM
              </Typography>
            </Box>
            <Typography id='compclock' data-testid='compclock'
             sx={{marginLeft: '-2px', letterSpacing: '-30px', lineHeight:.8, fontSize:230, fontWeight: 600, padding: 0, margin: 0}}>
               0
            </Typography>
            <Box ml={3} mt={1}>
              <Typography id='compminutes' data-testid='compminutes'
               sx={{letterSpacing: '-4px',lineHeight:.92, fontSize:150, fontWeight: 800, padding: 0, margin: 0}}>
                00
              </Typography>
              <Typography id='maindate' data-testid='maindate'
               sx={{ textTransform: 'none', lineHeight:1, fontSize:26, padding: 0, margin: 0}}>
                Day, 01/01
              </Typography>
            </Box>
           </Box>
        </Button>
        <Box mx={4} display='flex' justifyContent='flex-end'>
          <Button onClick={() => props.onComplete('digital1-color')}>Color</Button>
          <Button onClick={() => props.onComplete('close')}>Scheduler</Button>
        </Box>
      </>
) };



