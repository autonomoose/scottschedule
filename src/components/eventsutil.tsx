// events utilities and components
// exports default DisplayEvents,
//  also exports components CreateEvent, ModifyEvent
//  and data fetchEventsDB - full events
import React, { useEffect, useState } from 'react';
import { API } from 'aws-amplify';
import { useForm } from "react-hook-form";

import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import IconButton from '@mui/material/IconButton';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import Typography from '@mui/material/Typography';

import { listEventsFull } from '../graphql/queries';
import { mutAddEvents, mutAddRules, mutDelEvents } from '../graphql/mutations';

// -------------------------------------------------
interface CreateEventProps {
  onComplete?: (status: string) => void,
  open: boolean
}
export const CreateEvent = (props: CreateEventProps) => {
    // form states
    const { register, handleSubmit, reset, formState } = useForm({
        defaultValues: {
            name: '',
            descr: '',
        }
    });
    const { isDirty, errors } = formState;

    interface FormNewEvParms {
        name: string,
        descr: string,
    };
    const formNewEvSubmit = async (data: FormNewEvParms) => {
        console.log('form data', data);
        try {
            const xdata = {'input': {
                'etype': 'ev',
                'evnames': data.name+"!args",
                'descr': data.descr,
                }
            };
            const result = await API.graphql({query: mutAddEvents, variables: xdata});
            console.log('updated', result);
            if (props.onComplete) {
                props.onComplete(data.name);
            }
        } catch (result) {
            console.log('failed update', result);
        }
    };

    return(
      <Box display={(props.open)?'block': 'none'}>
      <Card style={{marginTop: '3px', maxWidth: 350, minWidth: 350, flex: '1 1', background: '#FAFAFA',
       boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <form key="newEv" onSubmit={handleSubmit(formNewEvSubmit)}>
          <Box display="flex" justifyContent="space-between" alignItems="baseline">
            <Typography variant='h6'>
              Add New Event
            </Typography>
          </Box>

          <Box><label>
            Name <input type="text" size={12} data-testid="nameInput"
             {...register('name', { required: true, pattern: /\S+/, maxLength:16 })}
             aria-invalid={errors.name ? "true" : "false"}
            />
          </label></Box>
          <Box><label> Description
            <input type="text" size={30} data-testid="descrInput"
             {...register('descr', { required: true, pattern: /\S+/, maxLength:30 })}
             aria-invalid={errors.descr ? "true" : "false"}
            />
          </label></Box>

          <Box mt={2} display='flex' justifyContent='flex-end'>
            <Button size="small" variant="outlined" onClick={() => reset()} disabled={!isDirty}>Reset</Button>
            <Button size="small" variant="contained" color="primary" type="submit" disabled={!isDirty}>Save</Button>
          </Box>

          </form>
        </Box>
      </Card></Box>
) };

// -------------------------------------------------
interface CreateRuleProps {
  evName: string,
  onComplete?: (status: string) => void,
  open: boolean
}
export const CreateRule = (props: CreateRuleProps) => {
    // form states
    const { register, handleSubmit, reset, formState, watch } = useForm({
        defaultValues: {
            cmd: 'begin',
            options: '',
            content: '',
        }
    });
    const { isDirty, errors } = formState;

    const formNewRuleCancel = async () => {
        if (props.onComplete) {
            props.onComplete('');
        }
    }
    interface FormNewRuleParms {
        cmd: string,
        options: string,
        content: string,
    };
    const formNewRuleSubmit = async (data: FormNewRuleParms) => {
        console.log('form data', data);
        try {
            let wkEvNames = props.evName+"!"+data.cmd;
            if (data.options !== '') {
                wkEvNames += ' ' + data.options;
            }
            const xdata = {'input': {
                'etype': 'ev',
                'evnames': wkEvNames,
                'rules': data.content,
                }
            };
            const result = await API.graphql({query: mutAddRules, variables: xdata});
            console.log('updated rules', result);
            if (props.onComplete) {
                props.onComplete(props.evName);
            }

        } catch (result) {
            console.log('failed update rules', result);
            if (props.onComplete) {
                props.onComplete('');
            }
        }
    };

    return(
      <Box display={(props.open)?'block': 'none'}>
      <Card style={{marginTop: '3px', maxWidth: 350, minWidth: 350, flex: '1 1', background: '#FAFAFA',
       boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <form key="newRule" onSubmit={handleSubmit(formNewRuleSubmit)}>
          <Box display="flex" justifyContent="space-between" alignItems="baseline">
            <Typography variant='h6'>
              Add New Rule
            </Typography>
          </Box>

          <Box display='flex'>
            <select {...register("cmd")}>
              <option value="begin">begin</option>
              <option value="option">option</option>
            </select>
            <input disabled={(watch('cmd') === 'begin')} data-testid="optionsInput" {...register('options')} />
          </Box>

          <Box>
            <textarea rows={2} cols={27} data-testid="contentRule"
             {...register('content', { required: true,})}
             aria-invalid={errors.content ? "true" : "false"}
            ></textarea>
          </Box>

          <Box mt={2} display='flex' justifyContent='flex-end'>
            <Button size="small" variant="outlined" color='error' onClick={() => formNewRuleCancel()}>Cancel</Button>
            <Button size="small" variant="outlined" onClick={() => reset()} disabled={!isDirty}>Reset</Button>
            <Button size="small" variant="contained" type="submit" disabled={!isDirty}>Save</Button>
          </Box>

          </form>
        </Box>
      </Card></Box>
) };

// -------------------------------------------------
interface ModifyEventProps {
  evid: string,
  tasks: iTask,
  onComplete?: (status: string) => void,
  open: boolean,
}
export const ModifyEvent = (props: ModifyEventProps) => {
    const allTasks = props.tasks;
    const evid = props.evid;

    const [evRule, setEvRule] = useState('');

    // form states
    const { register, handleSubmit, reset, formState } = useForm({
        defaultValues: {
            descr: '',
        }
    });
    const { isDirty, errors } = formState;

    useEffect(() => {
        const defaultValues = {
            descr: (allTasks[evid] && allTasks[evid].descr)? allTasks[evid].descr : '',
        }
        reset(defaultValues);
    }, [allTasks, evid] );

    interface FormModEvParms {
        descr: string,
    };
    const formModEvSubmit = async (data: FormModEvParms) => {
        console.log('modform data', data);
        try {
            const xdata = {'input': {
                'etype': 'ev',
                'evnames': evid+"!args",
                'descr': data.descr,
                }
            };
            const result = await API.graphql({query: mutAddEvents, variables: xdata});
            console.log('updated', result);
            if (props.onComplete) {
                props.onComplete(evid);
            }
        } catch (result) {
            console.log('failed update', result);
        }
    };
    interface FormDelEventParms {
        cmd: string,
    };
    const formDelEvent = async (data: FormDelEventParms) => {
        console.log('formDel parms', data);
        try {
            const xdata = {'input': {
                'etype': 'ev',
                'evnames': evid+"!"+data.cmd,
                }
            };
            console.log('deleting', xdata);
            const result = await API.graphql({query: mutDelEvents, variables: xdata});
            console.log('deleted', result);
            if (props.onComplete) {
                props.onComplete(evid);
            }
        } catch (result) {
            console.log('failed delete', result);
        }
    };
    const formCallback = (status: string) => {
        console.log("mod callback status", status);
        setEvRule('');
        if (props.onComplete && status !== '') {
            props.onComplete(status);
        }
    };

    return(
      <Box display={(props.open)?'block': 'none'}>
      <Card style={{marginTop: '3px', maxWidth: 350, minWidth: 350, flex: '1 1', background: '#FAFAFA',
       boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <form key="newEv" onSubmit={handleSubmit(formModEvSubmit)}>
          <Box display="flex" justifyContent="space-between" alignItems="baseline">
            <Typography variant='h6'>Modify Event</Typography>
          </Box>
          <Box display='flex' alignItems='center'>
            {evid}
            { (allTasks[evid] && allTasks[evid].schedRules.length === 0) &&
            <IconButton size='small' color='error' onClick={() => formDelEvent({'cmd': 'args'})}>X</IconButton>
            }
          </Box>

          <Box><label>
            <input type="text" size={30} data-testid="descrInput"
             {...register('descr', { required: true, pattern: /\S+/, maxLength:30 })}
             aria-invalid={errors.descr ? "true" : "false"}
            />
          </label></Box>

          <Box mt={2} display='flex' justifyContent='flex-end'>
            <Button size="small" variant="outlined" onClick={() => reset()} disabled={!isDirty}>Reset</Button>
            <Button size="small" variant="contained" color="primary" type="submit" disabled={!isDirty}>Save</Button>
          </Box>
          </form>
          <Box mb={1} display='flex' justifyContent='space-around'>
            <span>Rules ({(allTasks && allTasks[evid])? allTasks[evid].schedRules.length: 0})</span>
            <Button onClick={() => setEvRule(evid)}  size="small" variant="outlined" color="primary">New Rule</Button>
          </Box>
          {(allTasks && allTasks[evid]) &&
          <Box>
            <CreateRule evName={evRule} onComplete={formCallback} open={(evRule !== '')} />
            {
              allTasks[evid].schedRules.map((task: string) => {
                // get the cmd words out
                let words = task.split(' ');
                let cmd = words.shift() || 'begin';
                if (cmd === 'option') {
                    cmd = 'option ' +  words.shift() || '';
                }
                const displayRule = words.join(' ').split(',').join(', ');

              return(
                <Box key={`${evid}${cmd}`}  display='flex'>
                  <Box display='flex'>
                    <IconButton size='small' color='error' onClick={() => formDelEvent({cmd})}>X</IconButton>
                    {cmd}
                  </Box>
                  <Box mx={1} mb={1} px={1} sx={{ border: '1px solid grey' }} key={`${evid}${cmd}`}>
                    {displayRule}
                  </Box>
                </Box>
              )})
            }

          </Box>
          }

        </Box>
      </Card></Box>
) };

// -------------------------------------------------
interface DisplayEventProps {
  tasks: iTask,
  select?: (evid: string) => void,
}
const DisplayEvents = (props: DisplayEventProps) => {
    const allTasks = props.tasks;

    return(
      <List disablePadding dense sx={{marginLeft: '1em'}}>
        {
          Object.keys(allTasks).map((evid: string) => {
          return(
            <ListItem button key={evid} onClick={() => {props.select?.(evid);}}>
              {evid} - {allTasks[evid].descr}
            </ListItem>
           )})
        }
      </List>
) };

// -------------------------------------------------
export const fetchEventsDB = async (): Promise<iTask> => {
    try {
        const result: any = await API.graphql({query: listEventsFull})
        console.log("events loaded:", result.data.listEvents.items.length);

        const compactTasks = result.data.listEvents.items.reduce((resdict: iTask, item: iTaskDb) => {
            const evkeys = item.evnames.split('!');
            if (!resdict[evkeys[0]]) {
                resdict[evkeys[0]] = {descr: '', schedRules: []};
            }
            if (evkeys[1] === 'args') {
                resdict[evkeys[0]].descr = (item.descr)? item.descr: '';
            } else {
                resdict[evkeys[0]].schedRules.push(evkeys[1] + " " + item.rules);
            }
            return resdict;
        }, {});

        return(compactTasks);
    } catch (result) {
        console.log("got error", result);
        return({});
    }
};

export default DisplayEvents
