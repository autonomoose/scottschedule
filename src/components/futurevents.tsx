import React from 'react';

import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import Typography from '@mui/material/Typography';

interface iFutureDict {
    [evTstampKey: string]: iFutureEvent[],
}
interface DisplayFutureEventProps {
    item: iFutureEvent,
    descr?: string,
};
const DisplayFutureEvent = (props: DisplayFutureEventProps) => {
    const wkdate = new Date(props.item.evTstamp);
    const wkdescr = (props.descr)? props.descr: props.item.evTaskId;
    return (
      <div>
        {wkdate.toLocaleString('en-US', {hour: "2-digit", minute: "2-digit"})} - {wkdescr}
      </div>
)};
export default DisplayFutureEvent

interface DisplayFutureCardProps {
    evs: iFutureEvent[],
    tasks: iTask,
};
export const DisplayFutureCard = (props: DisplayFutureCardProps) => {
    return (
      <Card style={{marginTop: '3px', maxWidth: 432, minWidth: 404, flex: '1 1',
          background: '#F5F5E6', boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <Typography variant='h6'>
            Upcoming Events
          </Typography>
          { props.evs.slice(1).map(item => <DisplayFutureEvent
            key={`${item.evTstamp}:${item.evTaskId}`} item={item}
            descr={(props.tasks[item.evTaskId])? props.tasks[item.evTaskId].descr: 'system'}/>)
          }
        </Box>
      </Card>
)};


// future events
//
// derive the future events from schedule
//   get the start time figured out
//   get the Task rules that match
//   process rules into events dict w/ tstamp key
//   convert dict to sorted array w/ earliest events first
export const buildFutureEvents = (wkgroup: iSchedGroup, wksched: string, taskInfo: iTask, optInfo: iSchedOptions): iFutureEvs => {
    let wkEvents: iFutureEvent[] = [];
    let currdate = new Date(Date.now());
    if (optInfo['tomorrow']) {
        currdate.setHours(currdate.getHours() + 24);
        currdate.setHours(0);
        currdate.setMinutes(0);
        currdate.setSeconds(1);
    }

    // date object used for starting times,
    const tlangDate = (wkTlang: string | undefined, wkStart: Date) => {
        let retDate = new Date(wkStart.valueOf())
        if (wkTlang) {
            const dateParts = wkTlang.split(':');
            switch(dateParts.length) {
                case 1: // minutes only (offset only) or word
                    if (/^\+?\d+$/.test(wkTlang)) {
                        // numeric is minutes only
                        if (wkTlang[0] === '+') {
                            // offset
                            retDate.setMinutes(retDate.getMinutes() + Number(dateParts[0]));
                            retDate.setSeconds(0);
                        } else {
                            // constant
                            retDate.setMinutes(Number(dateParts[0]));
                            retDate.setSeconds(0);
                        }
                    } else {
                        switch(wkTlang) {
                            case 'now':
                                break;
                            default:
                                // console.log("unknown date - bad command", wkTlang);
                                break;
                        }
                    }
                    break;
                case 2: // hours and minutes (constant or offset)
                    if (wkTlang[0] === '+') {
                        // offset
                        retDate.setHours(retDate.getHours() + Number(dateParts[0].substring(1)));
                        retDate.setMinutes(retDate.getMinutes() + Number(dateParts[1]));
                        retDate.setSeconds(0);
                    } else {
                        // constant
                        retDate.setHours(Number(dateParts[0]));
                        retDate.setMinutes(Number(dateParts[1]));
                        retDate.setSeconds(0);
                    }
                    break;
                default:
                    // console.log("unknown date", wkTlang);
                    break;
            }
        }
        return retDate;
    };
    // reducer loop thru rules
    const rulesReduceToEvents = (outEvents: iFutureDict, taskRule: string) => {
        // globals startTlang, startDate
        const evTask = taskRule.split('.')[0];
        const ruleParts = taskRule.split('.')[1].split(',');
        var lastDate = new Date(startDate.valueOf());
        lastDate.setHours(0);
        lastDate.setMinutes(0);
        lastDate.setSeconds(0);

        // handle each rule as a set of compound statements comma separated
        for (const wkRule of ruleParts) {
            // get timeoffset info and generate possible event time

            let ruleWords = wkRule.split(' ');
            let tlangTimeWord = ruleWords.shift();
            if (!tlangTimeWord || tlangTimeWord === '') {
                continue;
            }

            let evTime = (tlangTimeWord.startsWith('++'))
                ? tlangDate(tlangTimeWord.slice(1), lastDate)
                : tlangDate(tlangTimeWord, startDate); // global startDate

            // possibly adjust evTime with orlater(+0:00)/orsooner(xx:xx)
            //     continue matching unless it fails
            let nextTlangWord = ruleWords.shift();
            while (nextTlangWord === 'or' && ruleWords.length > 0) {
                let nextTimeWord = ruleWords.shift();

                let nextEvTime = (nextTimeWord.startsWith('++'))
                    ? tlangDate(nextTimeWord.slice(1), lastDate)
                    : tlangDate(nextTimeWord, startDate); // global startDate

                // console.log("curr", evTime);
                // console.log("or ", nextTimeWord, nextEvTime);
                if (nextTimeWord[0] === '+') {
                    // or later
                    if (evTime > nextEvTime) {
                        // fails, no more or
                        break;
                    }
                    tlangTimeWord = nextTimeWord;
                    evTime = nextEvTime;
                } else {
                    // or sooner
                    if (evTime < nextEvTime) {
                        // fails, no more or
                        break;
                    }
                    tlangTimeWord = nextTimeWord;
                    evTime = nextEvTime;
                }
                nextTlangWord = ruleWords.shift();
            }

            // store event by timestamp for later extract in order
            if (typeof lastDate !== 'undefined' && lastDate.valueOf() === evTime.valueOf()) {
                console.log("repeat event not added");
            } else {
                let evVal = evTime.valueOf().toString();
                if (outEvents[evVal]) {
                    outEvents[evVal].push({evTstamp: evTime.valueOf(), evTaskId: evTask});
                } else {
                    outEvents[evVal] = [];
                    outEvents[evVal].push({evTstamp: evTime.valueOf(), evTaskId: evTask});
                }
                lastDate = evTime;
            }

            // deal with any * (repeat) instructions

        }
        return(outEvents);
    };
    // reducer loop thru tasks
    const tasksReduceToRules = (outRules: string[], wkTaskName: iSchedTask) => {
        // globals startTlang, startDate
        if (taskInfo[wkTaskName.evTaskId]) {
            // find matching rule - loop backward through rules to find first match
            let matchRule: string = '';
            const tasklist = taskInfo[wkTaskName.evTaskId].schedRules.slice().reverse();

            for (const wkRule of tasklist) {
                const ruleWords = wkRule.split(' ');
                switch(ruleWords[0]) {
                    case "begin":
                        // begin always matches
                        matchRule = ruleWords.slice(1).join(' ');
                        break;
                    case "option":
                        // option matches if all arguments are true
                        let options = ruleWords[1].split('+');
                        let matcher = true;
                        for (const wkOption of options) {
                            if (wkOption.startsWith('start:')) {
                                // todo ends with + should do date check
                                if (wkOption.slice(6) !== startTlang) { // global startTlang
                                    matcher = false;
                                    break;
                                }
                            } else if (!optInfo[wkOption]) {
                                matcher = false;
                                break;
                            }
                        }
                        if (matcher) {
                            matchRule = ruleWords.slice(2).join(' ');
                        }
                        break;
                    default:
                        console.log("unknown word", ruleWords[0]);
                    break;
                }
                if (matchRule !== '') {
                    break; // out of for loop
                }
            } // end of find matching rule

            // add matching rule to output
            if (matchRule !== '') {
                const outRule = wkTaskName.evTaskId + "." + matchRule
                outRules.push(outRule);
            }
        } else {
            console.log(wkTaskName.evTaskId, " task not found");
        }
        return outRules;
    }


    // find the schedule and possibly start time
    const schedParts = wksched.split('.');
    let schedList: iSchedule[] = [];
    if (wkgroup) {
        schedList = wkgroup.schedNames.filter(item => item.schedName === schedParts[0]);
    }
    if (schedList.length !== 1) {
        console.log("no unique schedule found ", schedList, schedParts[0]);
        return {evs:wkEvents};
    }
    let currSchedule = schedList[0];

    // find the starting time
    var startTlang = 'now';
    if (schedParts[1]) {
        startTlang = schedParts[1];
    } else if (currSchedule.begins) {
            startTlang = currSchedule.begins;
    }
    var startDate = tlangDate(startTlang, currdate);

    // get appropriate rules from tasklist
    let activeRules = [];

    if (currSchedule.schedTasks.length > 0) {
        activeRules = currSchedule.schedTasks.reduce(tasksReduceToRules, []);
    } else {
        // new
        activeRules.push("new.23:23");
        console.log("no tasks - added new rule");
    }

    // convert rules to future events dict (to preserve order on mult event per timestamp)
    const dictEvents = activeRules.reduce(rulesReduceToEvents, {});
    // convert dictionary to wkEvents: iFutureEvent[]
    for (const tmpTstamp of Object.keys(dictEvents).sort()) {
        for (const tmpEv of dictEvents[tmpTstamp]) {
            wkEvents.push(tmpEv);
        }
    }

    // copy currSchedule arguments that need to be on futureEvs
    // add another and this should be a set of fields processed by a loop
    let schedArgs: iFutureEvs = {evs: wkEvents, begins: startDate.valueOf()};
    if ('sound' in currSchedule) {
        schedArgs['sound'] = currSchedule['sound'];
    }
    if ('warn' in currSchedule) {
        schedArgs['warn'] = currSchedule['warn'];
    }
    return (schedArgs);
};
