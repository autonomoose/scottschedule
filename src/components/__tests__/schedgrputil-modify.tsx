import React from "react";
import { render, waitFor } from "@testing-library/react";
import userEvent from '@testing-library/user-event'

import { API } from 'aws-amplify';
import { ModifyGroup } from '../../components/schedgrputil';
jest.mock('aws-amplify');

const mockCallback = jest.fn();
const testSchedGroup = {
    name: 'testgrp',
    descr: 'test group',
    schedNames: [{
        begins: 'now',
        buttonName: 'test1',
        descr: 'test sched',
        schedName: 'testsched',
        schedTasks: [{evTaskId: 'testev'}],
        }]
};

const mytest = <ModifyGroup group='testgrp' groupSched={testSchedGroup} onComplete={mockCallback} open={true} />
const mySetup = () => {
    const utils = render(mytest);
    const resetButton = utils.getByRole('button', {name: /reset/i});
    const saveButton = utils.getByRole('button', {name: /save/i});
    const descrFld = utils.getByTestId('descrInput');
    const newButton = utils.getByRole('button', {name: /new schedule/i});

    return {
        ...utils,
        resetButton,
        saveButton,
        descrFld,
        newButton,
    }
};

describe("schedgrputil - modify", () => {
  it("renders snapshot correctly", () => {
    const {container} = render(mytest);

    expect(container.firstChild).toMatchSnapshot();
  });
  it("starts with buttons in correct status", () => {
    const utils = mySetup();

    expect(utils.resetButton).toBeDisabled();
    expect(utils.saveButton).toBeDisabled();
    expect(utils.newButton).toBeEnabled();
  });
  it("cancels with upper right x button", () => {
    const utils = mySetup();

    userEvent.click(utils.getByRole('button', {name: /x/i}));
    expect(mockCallback).toHaveBeenLastCalledWith('');
  });
  it("enables reset and save after descr modification", async () => {
    const utils = mySetup();

    userEvent.type(utils.descrFld, 'new desc');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    expect(utils.saveButton).toBeEnabled();
  });
  it("handles reset after descr modification", async () => {
    const utils = mySetup();

    userEvent.type(utils.descrFld, 'new descr');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    userEvent.click(utils.resetButton);
    await waitFor(() => {
      expect(utils.resetButton).toBeDisabled();
    });
  });

  it("handles graphql error on save", async () => {
    const consoleWarnFn = jest.spyOn(console, 'warn').mockImplementation(() => jest.fn());
    const prevAPIgraphql = API.graphql;
    API.graphql = jest.fn(() => Promise.reject('mockreject'));
    const utils = mySetup();

    userEvent.type(utils.descrFld, 'new desc');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    expect(utils.saveButton).toBeEnabled();
    userEvent.click(utils.saveButton);

    await waitFor(() => {
      expect(consoleWarnFn).toHaveBeenCalledTimes(1);
    });
    API.graphql = prevAPIgraphql;
    consoleWarnFn.mockRestore();
  });
  it("handles save after descr modification", async () => {
    const prevAPIgraphql = API.graphql;
    API.graphql = jest.fn(() => Promise.resolve({}));
    const utils = mySetup();

    userEvent.type(utils.descrFld, 'new desc');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    expect(utils.saveButton).toBeEnabled();
    userEvent.click(utils.saveButton);

    await waitFor(() => {
      expect(mockCallback).toHaveBeenLastCalledWith('testgrp');
    });
    API.graphql = prevAPIgraphql;
  });

});


