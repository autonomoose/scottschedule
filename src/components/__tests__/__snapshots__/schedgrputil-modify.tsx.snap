// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schedgrputil - modify renders snapshot correctly 1`] = `
<div
  class="MuiBox-root css-2t3rs8"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style="margin: 3px 0px 0px 0px; max-width: 350px; min-width: 350px; flex: 1 1; box-shadow: -5px 5px 12px #888888; border-radius: 0 0 5px 5px;"
  >
    <div
      class="MuiBox-root css-0"
    >
      <form>
        <div
          class="MuiBox-root css-1j6jckw"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Modify Group
          </h6>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            X
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-bqdthe"
        >
          testgrp
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
            data-testid="no-del"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1vqgfli-MuiSvgIcon-root"
              data-testid="DeleteForeverIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-o3707h"
        >
          <label>
            <input
              aria-invalid="false"
              data-testid="descrInput"
              name="descr"
              size="30"
              type="text"
            />
          </label>
        </div>
        <div
          class="MuiBox-root css-o3707h"
        />
        <div
          class="MuiBox-root css-1m55zth"
        >
          <button
            class="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Reset
          </button>
          <button
            class="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-11qr2p8-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
      <div
        class="MuiBox-root css-brm5ry"
      >
        <span>
          Schedules (
          1
          ) 
        </span>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-1k23hlb-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          New Schedule
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <ul
        class="MuiList-root MuiList-dense css-16vbaq4-MuiList-root"
      >
        <div
          class="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding MuiListItem-button css-qyxyi3-MuiButtonBase-root-MuiListItem-root"
          role="button"
          tabindex="0"
        >
          testsched
           - 
          test sched
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
  </div>
</div>
`;
